### DISCUSSION TITLE
How to change footer on the last page of a document that is multipage due to content

### DISCUSSION THREAD

**Discussion by SamuelScotford** [QuestPDF/QuestPDF#479] [General]
I have a document in which a large table is causing the creation of multiple pages (as intended) however on the final page I want to change the footer. How can I achieve this?

**Comment by luis-fss** [comment]
Indeed, it would be really cool to be able to configure a header and footer for the report in addition to those already available for pages.

**Comment by olsh** [comment]
> I have a document in which a large table is causing the creation of multiple pages (as intended) however on the final page I want to change the footer. How can I achieve this?

I believe that you can achieve that with the `Dynamic components` feature
https://www.questpdf.com/concepts/dynamic-components.html

Create a dynamic component like this
```csharp
public class FooterWithUniqueLastPage : IDynamicComponent<int>
{
    public int State { get; set; }

    public DynamicComponentComposeResult Compose(DynamicContext context)
    {
        var content = context.CreateElement(element =>
        {
            if (context.PageNumber == context.TotalPages)
            {
                element.Text("The last page");
            }
            else
            {
                element
                    .Text(x =>
                    {
                        x.CurrentPageNumber();
                        x.Span(" of ");
                        x.TotalPages();
                    });
            }
        });

        return new DynamicComponentComposeResult
        {
            Content = content,
            HasMoreContent = false
        };
    }
}

```

And use it
```csharp
page.Footer()
    .Dynamic(new FooterWithUniqueLastPage());
```

**Reply by harslannet** [reply]
Hi, If I create complex content, this check 'context.PageNumber == context.TotalPages' always returns 'false'!
But if I only add a Text element, it works correctly. 
What could be the reason?

**Reply by girlpunk** [reply]
Does the complex content have a fixed height, and if so is that the same height for both the final and previous pages?

**Reply by harslannet** [reply]
You asked a great question; yes, the complex content had a height, and removing it resolved the issue 😅😅  I can't believe I overlooked that! 😶‍🌫️ Thank you @girlpunk

**Reply by bterpperoni** [reply]
Link is dead, this is the new one : https://www.questpdf.com/concepts/code-patterns/dynamic-components.html @olsh

**Comment by naeemZz** [comment]
```c#
var content = context.CreateElement(container =>
{
    if(context.PageNumber != context.TotalPages)
    {
       container.Row() // code to show page number on footer
    }
    else
    {
     container.Row() // code to show last page footer
    }
}
```

Technically this should work, even `text.CurrentPageNumber()` are correct. Somehow context.TotalPages is showing some bizarrely large values. Below is a larger pseudo code:

```c#

private class DynamicFooter : IDynamicComponent<int>
{
    public int State { get; set; }
    private readonly BankStatement Model;

    public DynamicFooter(BankStatement model)
    {
        Model = model;
    }

    public DynamicComponentComposeResult Compose(DynamicContext context)
    {
    }    
}
```

**Reply by girlpunk** [reply]
As with the previous answer, any dynamic content that has large changes in the size required will create odd results given the nature of QuestPDF's two-pass rendering system.

**Reply by naeemZz** [reply]
There are no large changes in my case though. Maximum 2 pages while testing. Even then its not getting it correctly. I just want the footer of the last page to be different. Is there a better way to attain this?
Tx

**Comment by DemetrisMarkou** [comment]
Having the same issue with dynamic content. I get 5 pages instead of the two and when I debug the inital page number at the 1st pass is extremely large (more than a 1000). Somehow, questPDf corrects the page numbers at the 5 page mark. However, the original document is supposed to be 5. So, let me ask again. Is there an easier way to get a different footer on the last page?

**Reply by girlpunk** [reply]
Are you able to share your `IDynamicComponent` implementation that's causing problems?

**Reply by DemetrisMarkou** [reply]
[Quotation_XYZ006 (1).pdf](https://github.com/user-attachments/files/16062336/Quotation_XYZ006.1.pdf)


Certainly, and apologies for the late response. I was away for a business trip. I would appreciate any help with this. Also, I don't want to sound presumptuous but what if instead of relying on dynamic components there was an implementation of SkipOnce ShowOnce that works for footers, kinda similar to what there already is for Headers (maybe call it ShowLast). Here is my IDynamicComponent. I am also attaching the generated PDF. It's supposed to be 2 pages long. Thanks again for all your help.

`private class FooterComponent : IDynamicComponent
    {
        public DynamicComponentComposeResult Compose(DynamicContext context)
        {
            var currentPage = context.PageNumber;
            var totalPages = context.TotalPages;

            var content = context.CreateElement(element =>
            {
                element.Column(column =>
                {
                    // Page number for all pages
                    column.Item().AlignCenter().PaddingBottom(10).Text(text =>
                    {
                        text.CurrentPageNumber().FontColor("#040303");
                        text.Span("/").FontColor("#040303");
                        text.TotalPages().FontColor("#040303");
                    });

                    if (currentPage == totalPages)
                    {
                        // Last page footer
                        column.Item().PaddingBottom(10).Text("We will be happy to provide any further information you require.").FontColor("#151515");
                        column.Item().PaddingBottom(20).BorderTop(2).BorderColor("#A13333");
                        column.Item().Row(row =>
                        {
                            row.ConstantItem(400).AlignRight().Text("Signature:").FontSize(12);
                            row.ConstantItem(100).AlignRight().Image("wwwroot/Images/QuotationsApp/signature.png");
                        });
                        column.Item().Row(row =>
                        {
                            row.ConstantItem(400).Text("");
                            row.ConstantItem(100).Text("Panayiotis Pouris");
                        });
                    }

                    // "THANK YOU" text for all pages
                    column.Item().Background("#A13333").Height(20).AlignCenter().AlignMiddle().Text(text =>
                    {
                        text.DefaultTextStyle(x => x.FontSize(12).FontColor(Colors.White));
                        text.Span("THANK YOU FOR YOUR BUSINESS!");
                    });
                });
            });

            return new DynamicComponentComposeResult
            {
                Content = content,
                HasMoreContent = false
            };
        }
    }
}`

**Reply by girlpunk** [reply]
I think this may again be caused by the change in height between the preceding and final pages. Something like this may work as an alternative

```csharp
page.Content()
    .Column(c =>
    {
        c.Item()
            .ExpandVertical()
            .Decoration(d =>
            {
                d.Content()
                    .Text("Table of stuff goes here");
                d.After()
                    .AlignCenter()
                    .PaddingBottom(10)
                    .Text(text =>
                    {
                        text.CurrentPageNumber().FontColor("#040303");
                        text.Span("/").FontColor("#040303");
                        text.TotalPages().FontColor("#040303");
                    });
            });
        c.Item()
            .Column(column =>
            {
                // Last page footer
                column.Item()
                    .PaddingBottom(10)
                    .Text("We will be happy to provide any further information you require.")
                    .FontColor("#151515");
                column.Item()
                    .PaddingBottom(20)
                    .BorderTop(2)
                    .BorderColor("#A13333");
                column.Item()
                    .Row(row =>
                    {
                        row.ConstantItem(400)
                            .AlignRight()
                            .Text("Signature:")
                            .FontSize(12);
                        row.ConstantItem(100)
                            .AlignRight()
                            .Image("wwwroot/Images/QuotationsApp/signature.png");
                    });
                column.Item()
                    .Row(row =>
                    {
                        row.ConstantItem(400).Text("");
                        row.ConstantItem(100).Text("Panayiotis Pouris");
                    });
            });
    });

page.Footer()
    .Background("#A13333")
    .Height(20)
    .AlignCenter()
    .AlignMiddle()
    .Text(text =>
    {
        text.DefaultTextStyle(x => x.FontSize(12).FontColor(Colors.White));
        text.Span("THANK YOU FOR YOUR BUSINESS!");
    });
```

Alternatively, this may be caused by the page number variables being captured out of scope from their usage (i.e. with the `currentPage` and `totalPages` locals instead of referencing `context.PageNumber` and `context.TotalPages` directly)

**Reply by DemetrisMarkou** [reply]
[Quotation_DEM001.pdf](https://github.com/user-attachments/files/16123294/Quotation_DEM001.pdf)

It kinda works but it's placing the signature on a page all by itself, so it's not the desired results. I am not sure what I am doing wrong since I am new with QuestPDF but I would like to a) learn more (so if you can point me to more resources), b) make it work because I really like it.

**Reply by girlpunk** [reply]
If that's got you closer, could then try moving the `.ExpandVertical()` into the signature item, and then pushing the content of that item to the bottom of the container

**Reply by DemetrisMarkou** [reply]
I finally got it. 

I moved the page number to the `ComposeFooter` element (the footer) and then in the last page footer (`ComposeLastPageContent`) I create a single block of items for everything I wanted and pushed it to the bottom. 

Thank you so much for your help. I am posting my code here for anyone else who might struggle with the same issue.

`QuestPDF` is an awesome tool and I can't praise it high enough.

```c
//other stuff and methods

/* This is key here. `QuestPDF` renders the page content and then whenever that is done it renders 
       the `ComposeLastPageContent` */
 page.Content().Column(c =>
{
    c.Item().Element(container => ComposeMainContent(container, quotation, customer));
    c.Item().Element(ComposeLastPageContent); /* consider this an extension of the page.Content that loads after the 
                                                "main" content is done */
});

page.Footer().Element(ComposeFooter); // after the `page.Content` is done, `QuestPDF` renders the footer.
```
```c
       private void ComposeLastPageContent(IContainer container)
       {
           container.Column(column =>
           {

/* this is where I made the change. The whole signature thingie goes inside the same block (keeping the format 
I want with inner columns) and then extending the block to take the whole empty area. Align bottom 
pushes everything to the bottom of the container */

               column.Item().ExtendVertical().AlignBottom().Column(innerColumn =>
               {
                   innerColumn.Item().PaddingBottom(10).Text("We will be happy to provide any further information you require.").FontColor("#151515");

                   innerColumn.Item().PaddingBottom(20).BorderTop(2).BorderColor(red);

                   innerColumn.Item().Row(row =>
                   {
                       row.ConstantItem(400).AlignRight().Text("Signature:").FontSize(12);
                       row.ConstantItem(100).AlignRight().Image("wwwroot/Images/QuotationsApp/signature.png");
                   });

                   innerColumn.Item().Row(row =>
                   {
                       row.ConstantItem(400).Text("");
                       row.ConstantItem(100).Text("Panayiotis Pouris");
                   });
               });
           });
       }

       private void ComposeFooter(IContainer container)
       {
           container.Column(column =>
           {
               column.Item().AlignCenter().PaddingTop(5).PaddingBottom(10).Text(text =>
               {
                   text.CurrentPageNumber().FontColor(black);
                   text.Span(" / ").FontColor(black);
                   text.TotalPages().FontColor(black);
               });

               column.Item().Background(red).Height(20).AlignCenter().AlignMiddle().Text(text =>
               {
                   text.DefaultTextStyle(x => x.FontSize(12).FontColor(Colors.White));
                   text.Span("THANK YOU FOR YOUR BUSINESS!");
               });
           });
       }
```

**Comment by jhunexjun** [comment]
This was my code but somehow it's not working, it goes beyond the real total number of pages. So instead of 6 pages, it becomes 12 as you scroll down.
````
using QuestPDF.Elements;
using QuestPDF.Fluent;
using QuestPDF.Infrastructure;

namespace AutoGenReplenishment.Classes
{
    internal class LastPageFooter : IDynamicComponent<int>
    {
        public int State { get; set; }

        public DynamicComponentComposeResult Compose(DynamicContext context)
        {
            var content = context.CreateElement(element =>
            {
                if (context.PageNumber == context.TotalPages)
                {
                    element.Text("Last page here");
                }
                //else
                //{

                //}
            });

            return new DynamicComponentComposeResult
            {
                Content = content,
                HasMoreContent = false,
            };
        }
    }
}

````
Because of the error or bug or whatever, now this is my code:
```
using QuestPDF.Elements;
using QuestPDF.Fluent;
using QuestPDF.Infrastructure;
using System.ComponentModel;

namespace AutoGenReplenishment.Classes
{
    internal class LastPageFooter : IDynamicComponent<int>
    {
        public int State { get; set; }

        public DynamicComponentComposeResult Compose(DynamicContext context)
        {
            var content = context.CreateElement(element =>
            {
                if (context.PageNumber != context.TotalPages)
                {
                    element.Row(row =>
                    {
                        row.RelativeItem().Text("");
                    });
    }
                else
                {
                    element.Row(row =>
                    {
                        row.RelativeItem().Text("Equal");
                    }); // code to show last page footer
    }
            });

            return new DynamicComponentComposeResult
            {
                Content = content,
                HasMoreContent = false,
            };
        }
    }
}

```

**Comment by jhelmink** [comment]
Just to add my 2c to this topic in case any one else finds it.

The docs are clear for `page.Footer()` - [This container does not support paging capability](https://www.questpdf.com/api-reference/page/slots.html#main-slots).

I initially tried to use the same method of `.ShowIf(x => x.PageNumber == x.TotalPages)` but as described in this thread the `TotalPages` can't be known until after the first render.

Reading through this thread and thinking about it; having a Footer on the last page only, really is just bottom aligned content at the very end of your regular content (using `ExtendVertical` to use any remaining vertical space and `AlignBottom` to push to the bottom).

```
// At the very end of all content (aka final page) add a Footer
column.Item().ExtendVertical().AlignBottom().Element(c => GenerateFooterBlock(c));
```

**Reply by girlpunk** [reply]
> [This container does not support paging capability](https://www.questpdf.com/api-reference/page/slots.html#main-slots).

This refers to the fact that a footer cannot span multiple pages.

Your solution should work well if you would like a single footer only on the last page. Note that the original question being asked in this thread was for a footer that displayed on every page, with different content on the final page.

**Reply by jhelmink** [reply]
Thanks for the clarification, I think I got tunnel vision solving my particular problem.

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you to everyone who contributed to this discussion — your feedback helps us improve QuestPDF.

We are aware of the demand for this functionality and are currently exploring solutions along with their edge cases. We expect to release a fully supported version of this feature around Q3 2026.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
